### DISCUSSION TITLE
How does this compare against Valkey?

### DISCUSSION THREAD

**Discussion by re-thc** [dragonflydb/dragonfly#3854] [General]
Hi, most of the benchmarks are against Redis. Valkey now supports parallel processing and other features.

How does Dragonfly compare?

**Comment by kostasrim** [comment]
Hi @re-thc 

>  Valkey now supports parallel processing

From my understanding Valkey does not exactly support parallel processing. The commands you are running are still `single threaded` so once you saturate that you will hit a wall. From my current understanding parallel processing in Valkey, is not really parallel processing of commands but rather threads that deal with `IO`, `command parsing` etc. The idea is that the main thread executes commands while the rest of the IO threads parse, reply etc. That being said, this delegates a bunch of responsibility to the rest of the threads allowing the main thread to have time to execute more commands. However, this is not substitute for a shared memory (or a shared nothing) architecture where each of the threads can access any part of the redis dictionary (and actually execute the commands in parallel). That being said, there is still a hard limit on what you can actually parallize and what not as opposed to let's say dragonfly where each `thread` can execute commands. 

There is also the memory side. Df does not use Redis Dictionary to store key/values but rather something called a DashTable which effectively has less metadata waste (meaning that an entry in the dash table requires less bytes than the redis dictionary). Furthermore, our shared nothing architecture allows parallel execution due and our transaction framework does not require complicate locking strategies alleviating a lot of the bottlenecks associated with the traditional lock managers found in shared memory systems.

All in all my two cents are -- and without any actual benchmarks--  I would expect DF would offer way better performance and memory characteristics.

**Comment by alexted** [comment]
Less words, more deeds. Will you do a benchmark comparison?

There are already independent benchmarks of Valkey on many third-party sites, and it shows excellent results, not inferior to yours. Considering that Valkey supports all Redis commands and, most importantly, can scale horizontally in a cluster deployment (supported out of the box), I wonder how dragonfly will look against it.

**Comment by romange** [comment]
Sure, I will do these benchmarks soon. I am curious though could you please point me to these "independent benchmarks"?

**Reply by alexted** [reply]
I don't know if this article is available in English:
https://habr.com/en/articles/849264/ for example

**Reply by romange** [reply]
yeah, it's based on this original blog post:
https://valkey.io/blog/unlock-one-million-rps-part2/

**Reply by alexted** [reply]
The article summarises that their test results do not contradict the article you mention, but it is still a separate test.

**Reply by alexted** [reply]
Here are 2 more comparsions:
https://github.com/3scale-ops/valkey-performance-testing
https://openbenchmarking.org/test/pts/valkey

**Comment by Niennienzz** [comment]
Hey community friends @alexted @re-thc,

@romange  just published new benchmark results vs. Valkey.

Check out the numbers and details in the blog post ["Dragonfly vs. Valkey Benchmark: 4.5x Higher Throughput on Google Cloud" here](https://www.dragonflydb.io/blog/dragonfly-vs-valkey-benchmark-on-google-cloud).

As for clustering, we have related blog posts and services too. Please read more if you're interested:
- [Announcing Dragonfly Swarm](https://www.dragonflydb.io/blog/announcing-dragonfly-swarm-scaling-in-memory-data-without-limits)
- [A Preview of Dragonfly Cluster](https://www.dragonflydb.io/blog/a-preview-of-dragonfly-cluster)

**Comment by mlgitdev** [comment]
@romange / @Niennienzz 
Do you plan to do the same for AWS?

**Reply by Niennienzz** [reply]
Hey @mlgitdev, we will be releasing more benchmarks in the future.

It probably won't be exactly the same as the previous one but will definitely be AWS related.

**Reply by romange** [reply]
We just recently them. See here: https://www.dragonflydb.io/blog/value-benchmark-dragonfly-cloud-vs-elasticache
